src/System/include/BipedalLocomotion/System/AdvanceableRunner.h

```diff
@@ -0,0 +1,312 @@
+                    "{}.",
+                    errorPrefix,
+                    m_isTelemetryEnabled);
+    }
+
+    m_isInitialized = true;
+    return true;
+}
+
+template <class _Advanceable>
+bool AdvanceableRunner<_Advanceable>::setAdvanceable(std::unique_ptr<_Advanceable> advanceable)
+{
+    constexpr auto errorPrefix = "[AdvanceableRunner::setAdvanceable]";
+
+    if (advanceable == nullptr)
+    {
+        log()->error("{} The advanceable is not valid.", errorPrefix);
+        return false;
+    }
+
+    m_advanceable = std::move(advanceable);
+    return true;
+}
+
+template <class _Advanceable>
+bool AdvanceableRunner<_Advanceable>::setInputResource(std::shared_ptr<SharedResource<Input>> input)
+{
+    constexpr auto logPrefix = "[AdvanceableRunner::setInputResource]";
+
+    if (input == nullptr)
+    {
+        log()->error("{} The input is not valid.", logPrefix);
+        return false;
+    }
+
+    this->m_input = input;
+
+    return true;
+}
+
+template <class _Advanceable>
+bool AdvanceableRunner<_Advanceable>::setOutputResource(
+    std::shared_ptr<SharedResource<Output>> output)
+{
+    constexpr auto logPrefix = "[AdvanceableRunner::setOutputResource]";
+
+    if (output == nullptr)
+    {
+        log()->error("{} The input is not valid.", logPrefix);
+        return false;
+    }
+
+    this->m_output = output;
+
+    return true;
+}
+
+template <class _Advanceable> std::thread AdvanceableRunner<_Advanceable>::run()
+{
+    constexpr auto logPrefix = "[AdvanceableRunner::run]";
+
+    if (!m_isInitialized)
+    {
+        log()->warn("{} The AdvanceableRunner class is not initialized. An invalid thread will be "
+                    "returned",
+                    logPrefix);
+        assert(false && "[AdvanceableRunner::run] The thread is already running.");
+        return std::thread();
+    }
+
+    if (m_input == nullptr || m_output == nullptr)
+    {
+        log()->warn("{} The shared resources are not valid. An invalid thread will be returned.",
+                    logPrefix);
+        assert(false && "[AdvanceableRunner::run] The shared resources are not valid.");
+        return std::thread();
+    }
+
+    if (m_advanceable == nullptr)
+    {
+        log()->warn("{} The advanceable is not valid. An invalid thread will be returned.",
+                    logPrefix);
+        assert(false && "[AdvanceableRunner::run] The advanceable is not valid.");
+        return std::thread();
+    }
+
+    if (m_isRunning)
+    {
+        log()->warn("{} The thread is already running. An invalid thread will be returned.",
+                    logPrefix);
+        assert(false && "[AdvanceableRunner::run] The thread is already running.");
+        return std::thread();
+    }
+
+    // run the thread
+    m_isRunning = true;
+    auto function = [&]() -> bool {
+        while (this->m_isRunning)
+        {
+            const auto wakeUpTime = BipedalLocomotion::clock().now() + m_dT;
```

Comment:
Just FYI, this kind of implementation may lead to a drift in the execution (because the old `wakeUpTime` is not equal to `BipedalLocomotion::clock().now()`) that could problematic when doing data collection, see https://github.com/robotology/yarp/pull/2515 . I think that @isorrentino already experienced something similar in one of her issues.

Reply:
Do you have any suggestion to avoid this problem?

Reply:
I understood the problem! This thread is not periodic if there are deadline misses, right?

Reply:
```cpp
auto start = BipedalLocomotion::clock().now();
foo();
auto end = BipedalLocomotion::clock().now();

BipedalLocomotion::clock().sleepFor(m_dT - (end - start));
```

may solve the problem, right?

Reply:
or simply 
```
 wakeUpTime += m_dt:
```

But this may lead problem if the clock is reset

Reply:
> I understood the problem! This thread is not periodic if there are deadline misses, right?

Not only, in general you are not sure that the thread will wake up exactly when requested. The only thing that the API ensures is that it will wake up **after** the requested time. The only proper solution is the one that you suggest in the last comment, but as you correctly noted in that case you need to detect clock resets and account for them.

Reply:
If I'm not mistaken @Giulero proposed something in gazebo yarp plugin

Reply:
On detecting clock reset and handling them, you can also check https://github.com/robotology/yarp/pull/2494 and https://github.com/robotology/yarp/issues/800 .

Reply:
> If I'm not mistaken @Giulero proposed something in gazebo yarp plugins

Sorry, I sent my reply but I did not saw yours.

Reply:
Hi @traversaro here I tried to implement what you suggested while taking into account possible time reset https://github.com/dic-iit/bipedal-locomotion-framework/pull/272/commits/e91d6ad76906b293f1134663cc7f425d770a1b34

Reply:
Great, it seems ok!